File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        head = dummy
        l1 = list1
        l2 = list2
        while l1 or l2:
            if l1:
                n1 = l1.next
            if l2 :
                n2 = l2.next
            node = None
            if l1 and l2:
                if l1.val < l2.val:
                    node = l1
                    l1 = n1
                else:
                    node = l2
                    l2 = n2
            elif l1:
                node = l1
                l1 = n1
            else :
                node = l2
                l2 = n2
            head.next = node
            head = head.next
        return dummy.next

File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        auto* slow{head};
        auto* fast{head->next};
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* prev{nullptr};
        auto* cur{slow->next};
        slow->next = nullptr;
        while (cur) {
            auto* next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        auto* list1{head};
        auto* list2{prev};
        
        while (list1 && list2) {
            auto* next = list1->next;
            auto* next2 = list2->next;
            list1->next = list2;
            if (list2) list2->next = next;
            list1 = next;
            list2 = next2;
        }

    }
};
